jubilant/_pretty.py

```diff
@@ -0,0 +1,58 @@
+import dataclasses
+
+_MAX_VALUE = 150
+_INDENT = 2
+
+
+def _dump(value: object, indent='') -> str:
+    """Pretty-print a value with special cases for dataclasses, lists, and dicts.
+
+    If the value fits on a single (long) line, use that. Otherwise, split onto
+    multiple lines and indent each line.
+
+    For dataclasses, omit fields that are set to their default (or
+    default_factory return value).
+    """
+    sub_indent = indent + ' ' * _INDENT
+
+    if dataclasses.is_dataclass(value):
+        fields = []
+        for field in dataclasses.fields(value):
+            v = getattr(value, field.name)
+            if field.default is not dataclasses.MISSING and v == field.default:
+                continue
+            if field.default_factory is not dataclasses.MISSING and v == field.default_factory():
+                continue
+            v_str = _dump(v, sub_indent)
+            fields.append(f'{field.name}={v_str}')
+
+        class_name = value.__class__.__name__  # type: ignore
+        single_line = f'{class_name}({", ".join(fields)})'
+        if len(indent) + len(single_line) <= _MAX_VALUE:
+            return single_line
```

Comment:
Here, if the sub-`_dump()` returned a multi-line string, then we shouldn't ever return this level as a single string.

Actually I think that the code does work as is, albeit via a side-effect.

We're relying that the multi-line representation is always longer than a single-line representation. Which is true due to non-zero indent, I think.

If that didn't hold, we could end up with output like this:

```
TopLevelThing(field1=SubOne(
  sub_field1="something long",
  sub_field2="also long, causing a split"), field2=SubTwo(a=42))
```

If I read the code correctly, either document that, or propagate multi-lini-ness explicitly from sublevel to parent.